test/staking.ts

```diff
@@ -0,0 +1,354 @@
+import { StakingPoolFactory__factory, IRoleDefinition, RewardPool__factory } from "@energyweb/iam-contracts";
+import { StakingPoolFactory } from "@energyweb/iam-contracts/dist/ethers-v4/StakingPoolFactory";
+import { StakingPool } from "@energyweb/iam-contracts/dist/ethers-v4/StakingPool";
+import { EventFilter, Contract, Wallet, utils } from "ethers";
+import { Methods } from "@ew-did-registry/did";
+import { IAM, IStakable, Stakable, RegistrationTypes, setChainConfig } from "../src/iam-client-lib";
+import { claimManager, ensRegistry, rpcUrl, replenish, provider, deployer } from "./setup_contracts";
+import { root, rootOwner } from "./iam.test";
+import { mockJsonCodec, mockNats } from "./testUtils/mocks";
+import { BigNumber } from "ethers/utils";
+
+const { parseEther, namehash } = utils;
+export const waitFor = (filter: EventFilter, contract: Contract): Promise<any> => new Promise<any>((resolve) => {
+  contract.addListener(filter, (...args) => {
+    resolve(args);
+  });
+})
+  .then((args) => {
+    contract.removeAllListeners(filter);
+    return args;
+  });
+
+export const stakingTests = (): void => {
+  const serviceProvider = Wallet.createRandom().connect(provider);
+  const patron = Wallet.createRandom().connect(provider);
+  const patronDID = `did:${Methods.Erc1056}:${patron.address}`;
+
+  const orgName = "orgname";
+  const domain = `${orgName}.${root}`;
+  let serviceProviderIam: IAM & IStakable;
+  let patronIam: IAM & IStakable;
```

Comment:
Maybe a basic question here, but is it necessary to use interfaces here? If you had a `Staker` class, could you just do `IAM & Staker`? Removing the interfaces would be nice if we don't need the abstraction at this point, IMO

Reply:
Right. As there are no going to be other implementations of `IStakable` it should be abstract class

Reply:
I am agree that interface isn't the best way to typify staking feature, because interface assumes that there might be different implementations. But I can't figure out how `Staker` class to represent "addition of staking feature", because implementation of `Staker` will be based on IAM